sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala

```diff
@@ -224,8 +225,8 @@ abstract class BaseSessionStateBuilder(
    *
    * Note: this depends on `catalog` and `experimentalMethods` fields.
    */
-  protected def optimizer: Optimizer = {
-    new SparkOptimizer(catalog, experimentalMethods) {
+  protected def optimizer(catalogManager: CatalogManager): Optimizer = {
```

Comment:
We can create the `CatalogManager` in `BaseSessionStateBuilder` and pass it to analyzer/optimizer.

Reply:
OK. Changed.